pkg/install/quota.go

```diff
@@ -0,0 +1,83 @@
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	"github.com/Azure/ARO-RP/pkg/util/azureclient/mgmt/compute"
+)
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+//checkQuota checks usage quotas vs. resources required by cluster before cluster creation
+func (i *Installer) checkQuota(ctx context.Context) error {
+	//prepare usage client
+	spp := &i.doc.OpenShiftCluster.Properties.ServicePrincipalProfile
+	r, err := azure.ParseResourceID(i.doc.OpenShiftCluster.ID)
+	if err != nil {
+		return err
+	}
+	conf := auth.NewClientCredentialsConfig(spp.ClientID, string(spp.ClientSecret), spp.TenantID)
+	authorizer, err := conf.Authorizer()
+	if err != nil {
+		return err
+	}
+	usageClient := compute.NewUsageClient(r.SubscriptionID, authorizer)
+	return i.calculateAndCompareQuota(ctx, usageClient)
+}
+
+func (i *Installer) calculateAndCompareQuota(ctx context.Context, uc compute.UsageClient) error {
+	//master nodes are already included in the declaration below
+	requiredResources := map[string]int{
+		"cores":              24,
+		"virtualMachines":    3,
+		"standardDSv3Family": 3,
+		"PremiumDiskCount":   3,
+	}
+	if api.VMSizeStandardD8sV3 != i.doc.OpenShiftCluster.Properties.MasterProfile.VMSize {
+		//if master node type changed error out
+		return fmt.Errorf("Unexpected Master node VMSize %s, expected %s", i.doc.OpenShiftCluster.Properties.MasterProfile.VMSize, api.VMSizeStandardD8sV3)
+	}
+	//worker node resource calculation
+	for _, w := range i.doc.OpenShiftCluster.Properties.WorkerProfiles {
+		i.log.Infof("WorkerProfile VMSize %s\n", w.VMSize)
+		requiredResources["virtualMachines"] = requiredResources["virtualMachines"] + w.Count
+		requiredResources["PremiumDiskCount"] = requiredResources["PremiumDiskCount"] + w.Count
+		switch w.VMSize {
+		case api.VMSizeStandardD2sV3:
+			requiredResources["standardDSv3Family"] = requiredResources["standardDSv3Family"] + w.Count
+			requiredResources["cores"] = requiredResources["cores"] + (w.Count * 2)
+		case api.VMSizeStandardD4sV3:
+			requiredResources["standardDSv3Family"] = requiredResources["standardDSv3Family"] + w.Count
+			requiredResources["cores"] = requiredResources["cores"] + (w.Count * 4)
+		case api.VMSizeStandardD8sV3:
+			requiredResources["standardDSv3Family"] = requiredResources["standardDSv3Family"] + w.Count
+			requiredResources["cores"] = requiredResources["cores"] + (w.Count * 8)
+		default:
+			//will only happen if pkg/api verification allows new VMSizes
+			return fmt.Errorf("Unexpected Worker node VMSize %s", w.VMSize)
+		}
+
+	}
+
+	usages, err := uc.List(ctx, i.doc.OpenShiftCluster.Location)
+	if err != nil {
+		return err
+	}
+	//check requirements vs. usage
+	for _, usage := range usages {
+		required, present := requiredResources[*usage.Name.Value]
+		if present {
+			if int64(required) > (*usage.Limit - int64(*usage.CurrentValue)) {
+				return fmt.Errorf("Compute quota limit reached for \"%s\", needed %d, available %d (%d of %d used)",
```

Comment:
We want to propagate these errors to the users. Currently installer will not propagate these and return "Internal Server error". To propagate error you need to construt CloudError. Like here: https://github.com/Azure/ARO-RP/blob/46b3b9edbc562a5c251c1635a1c7fdc23b93dd86/pkg/util/dns/dns.go#L52 

and this code will be triggered and error returned to the user:
https://github.com/Azure/ARO-RP/blob/46b3b9edbc562a5c251c1635a1c7fdc23b93dd86/pkg/backend/openshiftcluster.go#L191-L201

Reply:
Not quite.  This code should be on the front end in the dynamic validator - it's like a preflight check.  It should return a CloudError like everything else in the dynamic validator.  On the back end there should be different code which can distinguish quota problems on the ARM deployments and also raise a CloudError - but that's in your VSTS story @mjudeikis.

Reply:
Customer facing error message guidance:
* start with a capital letter (you're doing this - just noting for completeness)
* end the error message with a `.`
* use single quotes as your error will be serialised to a JSON string
* simplify the message because it is going to have to be translated.  Suggestion for here: `Insufficient '%s' quota: need %d, current limit is %d.`